File: genetisk/src/main/java/no/asmund/genetisk/StartApp.java
package no.asmund.genetisk;

import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;
import javafx.stage.Stage;
import javafx.util.Duration;
import java.util.ArrayList;
import java.util.List;
import no.asmund.genetisk.Obstacle;

public class StartApp extends Application {
    public static final int HEIGHT = 600;
    public static final int WIDTH = 1200;
    public static int lifetime;
    private int lifecycle;
    public static int popNr;
    private GraphicsContext gc;
    private Timeline loop;
    private Population population;
    public static int targetX, targetY;
    public static List<Obstacle> obstacles = new ArrayList<>();

    @Override
    public void start(Stage primaryStage) {
        lifetime = 300;
        lifecycle = 0;
        popNr = 100;

        targetX = 1000;
        targetY = HEIGHT/2;

        obstacles.add(new Obstacle(WIDTH/2 - 50, HEIGHT/2 - 20, 100, 40));

        StackPane root = new StackPane();
        Scene scene = new Scene(root, WIDTH, HEIGHT);
        Canvas canvas = new Canvas(WIDTH, HEIGHT);

        gc = canvas.getGraphicsContext2D();
        root.getChildren().add(canvas);

        population = new Population(1, popNr);

        loop = new Timeline(new KeyFrame(Duration.millis(5), e -> draw()));
        loop.setCycleCount(Timeline.INDEFINITE);
        loop.play();
        primaryStage.setTitle("Genetisk Algoritme");
        primaryStage.setScene(scene);
        primaryStage.show();
    }

    public void draw() {
        if (lifecycle < lifetime) {
            repaint();
            lifecycle++;
        } else {
            gc.clearRect(0, 0, WIDTH, HEIGHT);
            lifecycle = 0;
            population.fitness();
            population.selection();
            population.reproduction();

        }
    }
    private void repaint() {
//        gc.clearRect(0, 0, WIDTH, HEIGHT);
        gc.setFill(Color.RED);
        gc.fillRect(targetX - 5, targetY - 5, 10, 10);

        for (Obstacle o : obstacles) {
            o.draw(gc);
        }

        gc.setFill(Color.BLACK);
        for (int i = 0; i < popNr; i++) {
            Creature r = population.population[i];
            r.run();
            gc.setFill(r.color);
            gc.fillOval(r.x - r.d / 2, r.y - r.d / 2, r.d, r.d);
        }
    }

}

File: genetisk/src/main/java/no/asmund/genetisk/StartApp.kt
package no.asmund.genetisk

import javafx.animation.KeyFrame
import javafx.animation.Timeline
import javafx.application.Application
import javafx.scene.Scene
import javafx.scene.canvas.Canvas
import javafx.scene.canvas.GraphicsContext
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color
import javafx.stage.Stage
import javafx.util.Duration

class StartApp : Application() {
    companion object {
        const val HEIGHT = 600
        const val WIDTH = 1200
        
        var lifetime: Int = 0
        var popNr: Int = 0
        var targetX: Int = 0
        var targetY: Int = 0
        
        @JvmStatic
        fun main(args: Array<String>) {
            launch(StartApp::class.java, *args)
        }
    }
    
    private var lifecycle = 0
    private lateinit var gc: GraphicsContext
    private lateinit var loop: Timeline
    private lateinit var population: Population
    
    override fun start(primaryStage: Stage) {
        lifetime = 300
        lifecycle = 0
        popNr = 100

        targetX = 1000
        targetY = HEIGHT / 2

        val root = StackPane()
        val scene = Scene(root, WIDTH.toDouble(), HEIGHT.toDouble())
        val canvas = Canvas(WIDTH.toDouble(), HEIGHT.toDouble())

        gc = canvas.graphicsContext2D
        root.children.add(canvas)

        population = Population(1000000f, popNr)

        loop = Timeline(
            KeyFrame(
                Duration.millis(5.0),
                { draw() }
            )
        )
        loop.cycleCount = Timeline.INDEFINITE
        loop.play()
        
        primaryStage.title = "Genetisk Algoritme"
        primaryStage.scene = scene
        primaryStage.show()
    }

    fun draw() {
        if (lifecycle < lifetime) {
            repaint()
            lifecycle++
        } else {
            gc.clearRect(0.0, 0.0, WIDTH.toDouble(), HEIGHT.toDouble())
            lifecycle = 0
            population.fitness()
            population.selection()
            population.reproduction()
        }
    }

    private fun repaint() {
        gc.fill = Color.RED
        gc.fillRect(
            (targetX - 5).toDouble(),
            (targetY - 5).toDouble(),
            10.0,
            10.0
        )
        
        gc.fill = Color.BLACK
        for (i in 0 until popNr) {
            val creature = population.population[i]
            creature?.run()
            creature?.color?.let { color ->
                gc.fill = color
                gc.fillOval(
                    (creature.x - creature.d / 2).toDouble(),
                    (creature.y - creature.d / 2).toDouble(),
                    creature.d.toDouble(),
                    creature.d.toDouble()
                )
            }
        }
    }
}